dns_server/src/crawler/core/mod.rs

```diff
@@ -0,0 +1,585 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Mintlayer network crawler core
+//!
+//! To keep things simple, the server will try to keep connections open to all reachable nodes.
+//! When a new outbound connection is made, a new DNS record is added (but only for nodes on default ports).
+//! When the connection is closed, the DNS record is removed.
+//! When a connection fails, the server increases the backoff time between connection attempts.
+//! If the number of failed connection attempts exceeds the limit, the address is removed from the list.
+//! Once-reachable and newer-reachable addresses have different connection failure limits
+//! (equivalent to about 1 month and about 1 hour, respectively).
+//!
+//! The core is deterministic (and therefore can't do IO operations itself).
+//! All input comes in the form of `CoreEvent` messages and all output comes in the form of `CoreCommand` messages.
+
+use std::{
+    collections::{btree_map::Entry, BTreeMap, BTreeSet},
+    marker::PhantomData,
+    net::{IpAddr, Ipv4Addr, Ipv6Addr},
+    sync::Arc,
+    time::Duration,
+};
+
+use common::chain::ChainConfig;
+use crypto::random::{seq::IteratorRandom, Rng};
+use logging::log;
+use p2p::{
+    error::P2pError,
+    message::{AnnounceAddrRequest, PeerManagerMessage, PingRequest, PingResponse},
+    net::{
+        default_backend::transport::TransportAddress,
+        types::{ConnectivityEvent, PeerInfo},
+        NetworkingService,
+    },
+    peer_manager::global_ip::IsGlobalIp,
+    types::peer_address::PeerAddress,
+};
+
+/// How many outbound connection attempts can be made per heartbeat
+const MAX_CONNECTS_PER_HEARTBEAT: usize = 25;
+
+/// When the server drops the unreachable node address. Used for negative caching.
+const PURGE_UNREACHABLE_TIME: Duration = Duration::from_secs(3600);
+
+/// When the server drops the unreachable node address that was once reachable. This should take about a month.
+/// Such a long time is useful if the server itself has prolonged connectivity problems.
+pub const PURGE_REACHABLE_FAIL_COUNT: u32 = 35;
+
+/// Address state transition
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AddressTransition {
+    Disconnected,
+    Connecting,
+    Connected,
+}
+
+/// Connection state of a potential node address (outbound only)
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AddressState {
```

Comment:
As discussed in the call, please consider making this a parameter in an AddressData method, `AddressData::transition_to()`

Reply:
Added to `AddressState`.